### ISSUE TITLE
How do you save BOTH the validation and training metrics in each batch or epoch??

### ISSUE THREAD

**Issue by ashwin-cognitiv** [dmlc/mxnet#6695]
## Environment info
Operating System: macOS

Package used (Python/R/Scala/Julia): Python

MXNet version: 0.10

Python version and distribution: Python 3.6.1

## What have you tried to solve it?
I wrote my own batch end callback function, which I passed into the fit method. 
I have an option to save the file for the specified period of batches or epochs. 
My method is like this: (In this example I'm saving the metrics every period of batches)

```python
def call(period):
     def callback(param):
          if param.nbatch % period == 0:
                name_value = param.eval_metric.get_name_value()
          # ... then I save the name_value list of tuples to a file to view the metrics later.

```

I'm successfully saving the metrics to a file, but I don't know if the metrics are metrics on the training data or the validation data. I'd like to be able to save both. But after looking through the docs and callback.py, I don't understand how you can save both the validation and training metrics in the same callback function. 

Has this issue been resolved or is there a workaround? - https://github.com/dmlc/mxnet/issues/798 

Can someone help me?
Thanks.

**Comment by chao1224** [comment]
I think by default, people only need to store the evaluation metrics on validation set, not on training set.

For this problem, I would can hack by manually iterate training and validation set.
```
    for batch in train_iter:
        labels = batch.label[0].asnumpy()
        mod.forward(batch, is_train=True)
        mod.update_metric(metric, batch.label)
        mod.backward()
        mod.update()
        pred = mod.get_outputs()
        # do evaluation based on labels and pred

    for batch in val_iter:
        labels = batch.label[0].asnumpy()
        mod.forward(batch, is_train=True)
        mod.update_metric(metric, batch.label)
        mod.backward()
        mod.update()
        pred = mod.get_outputs()
        # do evaluation based on labels and pred
```

And after all, remember to reset train_iter and val_iter by `train_iter.reset()` and `val_iter.reset()`.

**Comment by ashwin-cognitiv** [comment]
@chao1224 Thank you so much!

**Comment by ashwin-cognitiv** [comment]
@chao1224 : sorry, I just got another quick question. Since I'm collecting the metrics from get_name_value() (in metric.py class) inside each of those loops, before I iterate through the validation batch, do I need to call reset() on the metrics as well?

**Comment by chao1224** [comment]
@ashwin-cognitiv I'm not quite sure what do you mean by collecting the metrics. It would be great if you can put your codes here.

And I think finally you would use this function: `python/mxnet/model.score`, link is [here](https://github.com/dmlc/mxnet/blob/master/python/mxnet/model.py#L676). If so, you don't have to worry about that, because `reset=True` by default.
If you are not using this function, I would say it would be no harm to do `reset()` unless you need some online validation info.

**Comment by ashwin-cognitiv** [comment]
I have an object ('model_version') that model's some network options such as the iteration type (batch, or epoch):
```python
    def format_metrics(self, model, metrics, val_iter, test_iter = None):
        """
        :param model: mx.Module object
        :param metrics: mx.EvalMetric object
        :param val_iter: mx.io.CSVIter object
        :param test_iter: mx.io.CSVIter object (optional)
        :return: tuple (str, float) list
        """
        train_value = metrics.get_name_value()
        metrics.reset()

        for val_batch in val_iter:
            model.forward(val_batch, is_train = False)
            model.update_metric(metrics, val_batch.label)

        val_iter.reset()  
        valid_value = metrics.get_name_value()
        metrics.reset() - # necessary?

        test_types = []
        test_values = []

        if test_iter:
            for test_batch in test_iter:
                model.forward(test_batch, is_train = False)
                model.update_metric(metrics, test_iter.label)
            test_value = metrics.get_name_value()
            test_types, test_values = zip(*test_value)
            test_types = ['test-' + s for s in test_types]

        train_types, train_values = zip(*train_value)
        train_types = ['train-' + s for s in train_types]
        valid_types, valid_values = zip(*valid_value)
        valid_types = ['valid-' + s for s in valid_types]


for epoch in range(model_version.epochs):
     for batch_number, train_batch in enumerate(train_iter):
          model.forward(train_batch, is_train = True)
          model.update_metric(metrics, train_batch.label)
          model.backward()
          model.update()
          if model_version.iteration_type == 'batch':
               self.format_metrics(model, model_version.metrics, val_iter, test_iter)
     if model_version.checkpoint_iteration_type == IterationType.Epoch:
           self.save_weights(model.get_params()[0], epoch)
           self.save_checkpoint(model, model_version, epoch)
     train_iter.reset() # start the iterator from the first batch before starting a new epoch

**Comment by chao1224** [comment]
@ashwin-cognitiv I haven't tried it yet, as I always call this function `mod.score(val_iter, val_metrics, epoch=e)`, as the link I attached above.

And followed source code from [here](https://github.com/dmlc/mxnet/blob/master/python/mxnet/metric.py#L106-L109), what I would sugget is simply print the `num_inst` and `sum_metrics` to see if that's what you want.

BTW. My understanding of `self.num_inst` is the sample number that you want to evaluate on one data pass, which means you should reset it after each data pass. And if you are using the `mod.score()` function, it will create new [eval metrics](https://github.com/dmlc/mxnet/blob/master/python/mxnet/model.py#L693-L694), so you don't have to reset the metric.

**Comment by ashwin-cognitiv** [comment]
@chao1224 Thank you for the explanation! After I test it, if I see anything wrong I'll post here for future reference. Yeah, that makes sense. I can see why I should reset it, otherwise every call to my helper function will just keep accumulating the metrics and they will be averaged over the course of all my iterations.

Like you suggested, mod.score() seems like a great function for getting metrics.
